src/solvers/dg.jl

```diff
@@ -753,24 +773,97 @@ function compute_coefficients!(u, func, t, mesh::AbstractMesh{1}, equations, dg:
     end
 end
 
+# Standard function for computing coefficients of `func` (usually the initial condition)
 function compute_coefficients!(u, func, t, mesh::AbstractMesh{2}, equations, dg::DG,
                                cache)
     @threaded for element in eachelement(dg, cache)
         for j in eachnode(dg), i in eachnode(dg)
-            x_node = get_node_coords(cache.elements.node_coordinates, equations, dg, i,
-                                     j, element)
+            x_node = get_node_coords(cache.elements.node_coordinates, equations, dg,
+                                     i, j, element)
+            u_node = func(x_node, t, equations)
+            set_node_vars!(u, u_node, equations, dg, i, j, element)
+        end
+    end
+end
+# Specialized function for computing coefficients of `func`, here a discontinuous initial condition
+function compute_coefficients!(u, func::DiscontinuousInitialCondition, t,
+                               mesh::AbstractMesh{2}, equations, dg::DG, cache)
+    @threaded for element in eachelement(dg, cache)
+        for j in eachnode(dg), i in eachnode(dg)
+            x_node = get_node_coords(cache.elements.node_coordinates, equations, dg,
+                                     i, j, element)
+            # Shift the outer (i.e., ±1 on the reference element) nodes passed to the 
+            # `func` (usually the initial condition) inwards by the smallest amount possible,
+            # i.e., [-1 + ϵ, +1 - ϵ].
+            # This avoids steep gradients in elements if a discontinuity is right at a cell boundary,
+            # i.e., if the jump location `x_jump` is at the position of an interface which is shared by 
+            # the nodes x_{e-1}^{(i)} = x_{e}^{(1)}.
+            # In particular, this results in the typically desired behaviour for 
+            # initial conditions of the form
+            #           { u_1, if x <= x_jump
+            # u(x, t) = {
+            #           { u_2, if x > x_jump
+            if i == 1
+                x_node = SVector(nextfloat(x_node[1]), x_node[2])
+            elseif i == nnodes(dg)
+                x_node = SVector(prevfloat(x_node[1]), x_node[2])
+            end
+            if j == 1
+                x_node = SVector(x_node[1], nextfloat(x_node[2]))
+            elseif j == nnodes(dg)
+                x_node = SVector(x_node[1], prevfloat(x_node[2]))
+            end
```

Comment:
This assumes that the element lies in the coordinate direction, which may not be the case for unstructured and curved meshes in general, is it?

Reply:
True, we would need to check the direction of the face.